File: gunkim/today_problem/23_01_18/17352_여러분의다리가되어드리겠습니다.java
import java.util.*;
import java.io.*;

public class Main {


    private int n;
    private int[] parent;

    public void solution() throws Exception {
        BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
        n = Integer.parseInt(br.readLine());
        parent = new int[n + 1];

        for (int i = 1; i <= n; i++) {
            parent[i] = i;
        }
        StringTokenizer st;
        for (int i = 0; i < n - 2; i++) {
            st = new StringTokenizer(br.readLine());
            int start = Integer.parseInt(st.nextToken());
            int end = Integer.parseInt(st.nextToken());
            union(start, end);
        }

        StringBuilder sb = new StringBuilder();
        for (int i = 1; i <= n; i++) {//root node 하나와 고립된 node 하나를 찾음
            if(find(i) == i)
                sb.append(i).append(' ');
        }

        System.out.println(sb);
    }

    public void union(int x, int y) {
        x = find(x);
        y = find(y);
        if (x != y) {
            if (x < y) parent[y] = x;
            else parent[x] = y;
        }
    }

    public int find(int x) {
        if (parent[x] == x) {
            return x;
        }
        return parent[x] = find(parent[x]);
    }

    public static void main(String[] args) throws Exception {
        new Main().solution();
    }

}

File: LeeYooseok/BOJ_오늘의 문제/2023/2023-01-18/17352 여러분의 다리가 되어 드리겠습니다.py
# union-find 문제

N = int(input())
parent = [i for i in range(N + 1)]
def find(node):
    if parent[node] == node:
        return node

    parent[node] = find(parent[node])
    return parent[node]

def union(a, b):
    a = find(a)
    b = find(b)

    if a < b:
        parent[a] = b
    else:
        parent[b] = a


for _ in range(N - 2):
    u, v = map(int, input().split())
    union(u, v)

pivot = find(1)
for i in range(2, N+1):
    if pivot != find(i):
        print(pivot, i)
        exit()

